### PR TITLE
Cleanup: rename middle::ty::sty and its variants.

### PR DESCRIPTION
**Pull request by eefriedman**
Use camel-case naming, and use names which actually make sense in modern Rust.

### CODE REVIEW

**Comment by eddyb** [comment]
LGTM, but a could of names seem somewhat long: `TyReference` and `TyArrayOrSlice`.
Another thought is that the `Ty` prefix could be dropped *if* `Ty::Foo` could be made to work instead (right now it's not looking so great).

**Comment by eefriedman** [comment]
I can shorten `TyReference` to `TyRef` without much loss of clarity.

I don't really know how to shorten `TyArrayOrSlice`; `TyVec` like master uses is misleading, neither `TySquare` or `TyAOrS` is obvious, and I can't think of anything else.  I'm thinking it will get split into `TyArray` and `TySlice` in a followup, so maybe just leave it for now?

As for getting rid of the `Ty` prefix, it might be possible to rename `TypeVariants` to something shorter like `TyV`, sprinkle some `use rustc::middle::ty::TyV` into other modules, and write `TyV::Int` instead of `ty::TyInt`.  Not sure if that's worthwhile.

**Comment by nrc** [comment]
In general, I don't mind long names so much. TyRef is fine. I think TyArray would be fine instead of TyArrayOrSlice - it is close enough. Dropping the Ty prefix in favour of Ty:: would be nice. But I don't want to see TyV at all - that's indecipherable. (In general, we have a history of short and somewhat bad names in the compiler and we've been moving towards longer, more understandable names. Obviously there are specific places where concision is still warranted though).

File: src/librustc/middle/ty.rs [hunk 1]
```diff
@@ -1335,53 +1336,101 @@ pub enum BoundRegion {
...
 // AST structure in libsyntax/ast.rs as well.
 #[derive(Clone, PartialEq, Eq, Hash, Debug)]
-pub enum sty<'tcx> {
-    ty_bool,
-    ty_char,
-    ty_int(ast::IntTy),
-    ty_uint(ast::UintTy),
-    ty_float(ast::FloatTy),
-    /// Substs here, possibly against intuition, *may* contain `ty_param`s.
+pub enum TypeVariants<'tcx> {
```
Comment (nrc) [inline src/librustc/middle/ty.rs]: maybe `Types` rather than `TyV` or `TypeVariants`?

File: src/librustc/middle/ty.rs [hunk 2]
```diff
@@ -1335,53 +1336,101 @@ pub enum BoundRegion {
...
+    /// `<T as Trait<..>>::N`.
+    TyProjection(ProjectionTy<'tcx>),
 
-    ty_tup(Vec<Ty<'tcx>>),
+    /// A type parameter; for example, `T` in `fn f<T>(x: T) {}
+    TyParam(ParamTy),
 
-    ty_projection(ProjectionTy<'tcx>),
-    ty_param(ParamTy), // type parameter
+    /// A placeholder type used during type-checking.
```
Comment (nrc) [inline src/librustc/middle/ty.rs]: I wouldn't call this a placeholder. We use "type variable" elsewhere in the compiler.

**Comment by nrc** [comment]
This is a great PR, I've wanted to rename these things for a while. Thanks for making the changes. If you could at least change TyArrayOrSlice and TyReference we could land this. You could try changing to use the prefix, or we could land this and do that as a follow up - I realise this PR will rot quite quickly.

**Comment by eefriedman** [comment]
Rebased, and amended to use TyRef and TyArray.

I think I'll experiment with the prefixes later; there's already a variant named "Types" in librustc/middle/infer/mod.rs.

**Comment by eefriedman** [comment]
I didn't rebase correctly; sorry about the churn.  Amended with fix.

File: src/librustc/middle/ty.rs (hunk 1 above)
Comment (eddyb) [inline src/librustc/middle/ty.rs]: I think @nikomatsakis also mentioned `TyKind` a long time ago (when I considered making this cleanup).

**Comment by eefriedman** [comment]
Updated again; I thought I had built everything, but I somehow missed librustdoc.

**Pull request merged**